nservicebus/sagas/saga-not-found.md

```diff
@@ -29,3 +29,67 @@ The ability to provide an implementation for `IHandleSagaNotFound` is especially
 For example, consider a saga that is used for managing the registration process on the website. After a customer registers, they receive an email with a confirmation link. The system will wait for confirmation for a specific period of time, e.g. 24 hours. If the user doesn't click the link within 24 hours, their data is removed from the system and saga is completed. However, they might decide to click the confirmation link a few days later. In this case, the related saga instance can't be found and an exception will be thrown. By implementing `IHandleSagaNotFound` it is possible to handle the situation differently, e.g. redirect the user to the registration website and ask them to fill out the form again.
 
 The implementation of `IHandleSagaNotFound` should be driven by the business requirements for a specific situation. In some cases the message might be ignored; in others, it might be useful to track whenever that situation happens (e.g. by logging or sending another message). In still other cases, it might make sense to perform a custom compensating action.
+
+## Troubleshooting
+
+Saga not found can occur when:
+
+1. The saga instance does not exist yet
+2. The saga instance already is removed
+
+Sometimes its not always that obvious when saga state will exist or be gone due to race conditions, false assumptions
+
+### Out of order
+
+If the design assumes message A will create the saga and B updates the saga. If message B is received thus processed before A the saga will not yet exist and will be discarded.
+
+Symptoms:
+
+- Saga are created, expected to be completed but this does not always happen.
+
+Mitigation:
+
+- Create the saga for both message A and B.
+
+### Concurrency
+
+Messages might be processed concurrently. If the saga is started by messages 1 but that did not yet complete thus create the saga state and message 2 is processed which depends on the existance of the saga state this message will be discarded with saga not found.
+
+Symptoms:
+
+- Scatter/gather sagas behave unexpectedly and show saga not found.
+
+Mitigation:
+
+- Reduce processing concurrency to 1 to achieve sequential processing. This only works that only a single endpoint instance active.
```

Comment:
or use pessimistic locking where available. Or redesign the saga to make so that all message types can start it

Reply:
@mauroservienti pessimistic locking 
```suggestion
- Reduce processing concurrency to 1 to achieve sequential processing. This only works that only a single endpoint instance active.
- Use pessimistic locking on the saga persister, this only works if a storage is configured that supports pessimistic locking.
```

Reply:
@timbussmann Based on the docs only Azure and non-durable it seems. RavenDB doesn't support pessimistic locking but we use a "lease" construct. So the effect it the same as long as the saga will not be invoked longer then 60 seconds.

But this applies to pre v7 too for which basically most persisters didn't support pessimistic locking. As least not SQLP and RavenDB.